### PR TITLE
Implemented RAG Chat bot feature

### PR DESCRIPTION
**Pull request by MukuFlash03**
Got the RAG based chat bot to work.
Focussed on changes relating to usage of transcript as the context for the bot to give answers from.

Added two tutorial python Jupyter notebooks: tut-1 and tut-2.

---------

In terms of the initial setup:
- Had to install ffmpeg on macOS; might have to do it similarly for Windows.
- Added some more dependencies to requirements.txt like deepgram.
- Updated all environment variables I've used in Google doc.

---------

Some changes:
1. title
- Some titles had punctuation symbols like &, emojis and the open(audio_file) was failing for such titles.
- Hence, changed to use a common static title for all video files being downloaded as audio files.

---------

2. ragChain.py
- Tut-1 and deepgram_embed.py is the basis for the ragChain.py file containing the RAG code.
- Major steps include:
- - create_chunks(): Converts text to documents in chunks.
- - create_vector_store(docs): Returns vector_store creating using docs, embeddings.
- - query_vector_store(vector_store, query): Just used for querying; can be skipped.
- - build_rag_chain(vector_store): Builds rag_chain, retriever
- - query_rag_chain(rag_chain, retriever, query): Returns answers to questions / chat messages based on current context.

---------

3. Fixed openai version issue with ChatCompletion.create()
- This was deprecated in latest version and correct usage include chat.completions.create()

---------

4. Storing transcript in streamlit session_state
- The app crashed on sending a successive chat message after initial message as the transcript value wasn't being saved since the server was sending request to the backend and didn't load the get_transcript() everytime.
- Hence, storing it as a streamlit session_state variable, so that it's available across files and for the entirety of the current browser session or until a new video link is uploaded.

### CODE REVIEW

File: ragChain.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+rag_chain: any
+retriever: any
+
+def create_chunks():
+    if not st.session_state.transcript:
+            return []
+    
+    # Using CharacterTextSplitter for using transcript directly
+    # text_splitter = CharacterTextSplitter(chunk_size=200, chunk_overlap=20)
```
Comment (marmikpatel19) [inline ragChain.py]: we can probably just get rid of unused code at the point

File: requirements.txt [hunk 2]
```diff
@@ -2,9 +2,15 @@ requests
 pymongo
 together
 streamlit==1.24.1
-openai==0.27.8
+openai==1.31.1
```
Comment (marmikpatel19) [inline requirements.txt:5]: nice! haha

**Review by marmikpatel19 — APPROVED** [review]
Thanks for getting this to work Mukul! The tutorials are nice touch as well :)

File: ragChain.py (hunk 1 above)
Comment (MukuFlash03) [reply ragChain.py]: Sure, will do, removed in latest commit.
Had just found two different approaches to pass in the transcript to the splitter - 1) document and 2) text. Both required different text_splitter mechanisms.

File: requirements.txt (hunk 2 above)
Comment (MukuFlash03) [reply requirements.txt:5]: Yes! 
Found that the way we were doing it was deprecated now, so using the [latest API documentation](https://platform.openai.com/docs/api-reference/completions/create?lang=python) found here for creating the chat completion.

**Pull request merged**